packages/sources/coinbase-prime/src/transport/balance.ts

```diff
@@ -0,0 +1,119 @@
+    total: string // Returns total in fiat amount
+    holds: string
+  }
+  vault_balances: {
+    total: string // Returns total in fiat amount
+    holds: string
+  }
+}
+
+export type HttpTransportTypes = BaseEndpointTypes & {
+  Provider: {
+    RequestBody: never
+    ResponseBody: ResponseSchema
+  }
+}
+export const httpTransport = new HttpTransport<HttpTransportTypes>({
+  prepareRequests: (params, config) => {
+    return params.map((param) => {
+      const timestamp = Math.floor(Date.now() / 1000)
+      const method = 'GET'
+      const path = `/v1/portfolios/${param.portfolio}/balances`
+      const message = `${timestamp}${method}${path}`
+      const signature = sign(message, config.SIGNING_KEY)
+      return {
+        params: [param],
+        request: {
+          baseURL: config.API_ENDPOINT,
+          url: path,
+          headers: {
+            'X-CB-ACCESS-KEY': config.ACCESS_KEY,
+            'X-CB-ACCESS-PASSPHRASE': config.PASSPHRASE,
+            'X-CB-ACCESS-SIGNATURE': signature,
+            'X-CB-ACCESS-TIMESTAMP': timestamp,
+            'Content-Type': 'application/json',
+          },
+          params: {
+            symbols: param.symbol.toUpperCase(),
+            balance_type: `${param.type.toUpperCase()}_BALANCES`,
+          },
+        },
+      }
+    })
+  },
+  parseResponse: (params, response) => {
+    return params.map((param) => {
+      if (!response.data) {
+        return {
+          params: param,
+          response: {
+            errorMessage: `The data provider did not return data for Portfolio: ${param.portfolio}, Balance Type: ${param.type}, Symbol: ${param.symbol}`,
+            statusCode: 502,
+          },
+        }
+      }
+
+      if (!response.data.balances) {
+        return {
+          params: param,
+          response: {
+            errorMessage: `The data provider response does not contain a balances list for Portfolio: ${param.portfolio}, Balance Type: ${param.type}, Symbol: ${param.symbol}`,
+            statusCode: 502,
+          },
+        }
+      }
+
+      // The adapter only supports querying one asset at a time so the balances list should only contain 1 element
+      if (response.data.balances.length !== 1) {
+        return {
+          params: param,
+          response: {
+            errorMessage: `The data provider response does not contain exactly one element in the balances list for Portfolio: ${param.portfolio}, Balance Type: ${param.type}, Symbol: ${param.symbol}`,
+            statusCode: 502,
+          },
+        }
+      }
+
+      const result = Number(response.data.balances[0].amount)
+      if (isNaN(result)) {
+        return {
+          params: param,
+          response: {
+            errorMessage: `The data provider returned non-numeric balance: ${response.data.balances[0].amount}`,
+            statusCode: 502,
+          },
+        }
+      }
+      return {
+        params: param,
+        response: {
+          result: result,
```

Comment:
i'm unclear when we should be using SingleNumberResponse vs something like PoRBalanceResponse, any tips?

Reply:
The latter is what we'd use for a balance adapter part of the `proof-of-reserves` composite stack. Most of those PoR types were made for adapters used with the `proof-of-reserves` composite adapter. They help standardize the input and output the composite expects. With adapters like these that use a 3rd party API, we'll usually use `SingleNumberResponse` like a price feed adapter since they're standalone.